libs/src/service-selection/ServiceSelection.js

```diff
@@ -0,0 +1,223 @@
+   * once. Tweaking this value may impact browser performance
+   * @param {number} config.requestTimeout the timeout at which to give up on a service
+   * @param {number} config.unhealthyTTL the point at which the unhealthy services are freed so they
+   * may be tried again (re-requested)
+   * @param {number} config.backupsTTL the point at which backup services are freed so they may be
+   * tried again (re-requested)
+   */
+  constructor ({
+    whitelist,
+    getServices,
+    maxConcurrentRequests = 6,
+    requestTimeout = 30 * 1000, // 30s
+    unhealthyTTL = 60 * 60 * 1000, // 1 hour
+    backupsTTL = 2 * 60 * 1000 // 2 min
+  }) {
+    this.whitelist = whitelist
+    this.getServices = getServices
+    this.maxConcurrentRequests = maxConcurrentRequests
+    this.requestTimeout = requestTimeout
+    this.unhealthyTTL = unhealthyTTL
+    this.backupsTTL = backupsTTL
+
+    // Truly "unhealthy" services. Should not ever be picked.
+    this.unhealthy = new Set([])
+
+    // Selectable services but not optimal. Will be picked as a last resort.
+    this.backups = {}
+
+    // Total number of services attempted
+    this.totalAttempts = 0
+  }
+
+  async select () {
+    // If a short circuit is provided, take it. Don't check it, just use it.
+    const shortcircuit = this.shortcircuit()
+    if (shortcircuit) return shortcircuit
+
+    // Get all the services
+    let services = await this.getServices()
+
+    // If a whitelist is provided, filter down to it
+    if (this.whitelist) {
+      services = this.filterToWhitelist(services)
+    }
+
+    // Filter out anything we know is already unhealthy
+    const filteredServices = this.filterOutKnownUnhealthy(services)
+
+    // Randomly sample a "round" to test
+    const round = this.getSelectionRound(filteredServices)
+
+    this.totalAttempts += round.length
+
+    // If there are no services left to try, either pick a backup or return null
+    if (filteredServices.length === 0) {
+      if (Object.keys(this.backups).length > 0) {
+        // Some backup exists
+        return this.selectFromBackups()
+      } else {
+        // Nothing could be found that was healthy.
+        // Reset everything we know so that we might try again.
+        this.unhealthy = []
+        this.backups = {}
+        return null
+      }
+    }
+
+    // Race this "round" of services, getting the best and ones that errored
+    // Note: ones that did not error or were not the best just get canceled so
+    // we don't really know anything about them at this point.
+    const { best, errored } = await this.race(round)
+
+    // Mark all the errored ones as unhealthy
+    errored.forEach(e => this.addUnhealthy(e))
+
+    // Trigger a cleanup event for all of the unhealthy and backup services,
+    // so they can get retried in the future
+    this.triggerCleanup()
+
+    // Recursively try this selection function if we didn't find something
+    if (!best) {
+      return this.select()
```

Comment:
im thinking that if for some reason all the services are down/some error occurs in the selection process, that this recursive call would go into an infinite loop. maybe bring back a global count to ensure that doesn't happen?

Reply:
this doesn't actually explicitly handle retries, so that's not possible. We'll try every service eventually if we don't find anything. Go through backups and if we still don't find anything, we just return null